sharedbuffers/mapped_struct.py

```diff
@@ -342,6 +342,194 @@ def unpack_from(cls, buf, offs, idmap = None):
         key, values = mapped_list.unpack_from(buf, offs, idmap)
         return cls(zip(key, values))
 
+
+def _upsize(x):
+    "Round up X to the next power of 2"
+    for p2 in (1, 2, 4, 8, 16, 32, 64):
+        x |= x >> p2
+    return x + 1
+
+def _add_kv(indexes, kvpairs, key, val):
+    l2 = len(indexes) / 2
+    hval = hash(key)
+    index = hval % l2
+    nprobe = 1
+
+    while True:
+        i2 = index + index
+        if indexes[i2] == -1:
+            indexes[i2] = len(kvpairs)
+            indexes[i2 + 1] = hval
+            kvpairs.extend((key, val))
+            return
+
+        index = (index + nprobe) % l2
+        nprobe += 1
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(
+        index_table=tuple,
+        kvpairs=proxied_list,
+        mask=int,
+        nelems=int)
+
+    def __init__(self, mask, nelems, index_table, kvpairs):
+        self.mask = mask
+        self.nelems = nelems
+        self.index_table = index_table
+        self.kvpairs = kvpairs
```

Comment:
For this to be a proxy, you can't have all the items as python objects. You need to only store a pointer to the position in the buffer where the data is, and then implement all getters by parsing the on-disk structure.

Reply:
I'm delegating this to proxied_list. Doesn't it handle that ?

Reply:
You built a tuple with the index.

The kvpairs proxied list contains tuples (regular tuples) inside it, when you access an item you'll be building that tuple.

You've got lots of holes in there.

Reply:
Huh ? The kvpairs proxied_list is flat. It doesn't contain tuples unless the dict itself has tuples as keys or values.

Reply:
Ah, yes, you're using extend. Still moot, you've copied the index into a tuple.

Just assume you've got a several-GB dict you've mmaped from a file here, and you still want instant unpack and instant access to an item.

Given this schema:

``` python
class Node:
    __slot_types__ = {
        'branches' : proxied_dict
    }
    __schema__ = Schema.from_typed_slots(__slot_types__)
```

Build a really huge proxied_dict and compare the performance of doing:

``` python
with open('somefile', 'rb') as f:
    f.seek(0, os.SEEK_END)
    size = f.tell()
    buf = mmap.mmap(f.fileno(), size, access = mmap.ACCESS_READ)
    node = node.__schema__.unpack_from(buf)
    node.branches[key]
```

It should be really fast, independent of file size, and use almost no memory.